src/mixins/java/org/spongepowered/common/mixin/core/world/food/FoodDataMixin.java

```diff
@@ -0,0 +1,128 @@
+    private void impl$tickDrainFoodLevel(final FoodData self, final int value) {
+        this.foodLevel = this.fireEventAndGetValue(Keys.FOOD_LEVEL, this.foodLevel, value);
+    }
+
+    @Redirect(method = "eat(IF)V", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;saturationLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$eatSetSaturationLevel(final FoodData self, final float value) {
+        this.saturationLevel = this.fireEventAndGetValue(Keys.SATURATION, (double) this.saturationLevel, (double) value).floatValue();
+    }
+
+    @Redirect(method = "tick", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;saturationLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$tickDrainSaturationLevel(final FoodData self, final float value) {
+        this.saturationLevel = this.fireEventAndGetValue(Keys.SATURATION, (double) this.saturationLevel, (double) value).floatValue();
+    }
+
+    @Redirect(method = "addExhaustion", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;exhaustionLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$addExhaustion(final FoodData self, final float value) {
+        this.exhaustionLevel = this.fireEventAndGetValue(Keys.EXHAUSTION, (double) this.exhaustionLevel, (double) value).floatValue();
+    }
+
+    @Redirect(method = "tick", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;exhaustionLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$tickDrainExhaustion(final FoodData self, final float value) {
+        this.exhaustionLevel = this.fireEventAndGetValue(Keys.EXHAUSTION, (double) this.exhaustionLevel, (double) value).floatValue();
+    }
+
+    private <E> E fireEventAndGetValue(final Key<? extends Value<E>> key, final E currentValue, final E value) {
+        if (!ShouldFire.CHANGE_DATA_HOLDER_EVENT_VALUE_CHANGE || Objects.equals(currentValue, value)) {
+            return currentValue;
+        }
+
+        final DataTransactionResult transaction = DataTransactionResult.builder()
+                .replace(Value.immutableOf(key, currentValue))
+                .success(Value.immutableOf(key, value))
+                .result(DataTransactionResult.Type.SUCCESS)
+                .build();
+
+        final ChangeDataHolderEvent.ValueChange
+                event =
+                SpongeEventFactory.createChangeDataHolderEventValueChange(PhaseTracker.getCauseStackManager().currentCause(), transaction, (DataHolder.Mutable) this.impl$player);
+
+        Sponge.eventManager().post(event);
+
+        if (event.isCancelled()) {
+            return currentValue;
+        }
+
+        for (final Value.Immutable<?> immutableValue : event.endResult().successfulData()) {
+            if (immutableValue.key() == key) {
+                return (E) immutableValue.get();
+            }
+        }
+
+        return currentValue;
+    }
```

Comment:
@gabizou Just want to check with you to make sure this is the right way to do `DataHolderEvent.ValueChange` events for specific keys. If it looks okay, this is good to go, if it doesn't, let us know how to sort it (@aromaa I can take over if you're unable to complete this, I know you said you have things coming up!)